sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectExecutionManager.scala

```diff
@@ -77,26 +77,30 @@ private[connect] class SparkConnectExecutionManager() extends Logging {
         request.getSessionId,
         previousSessionId)
     val executeHolder = new ExecuteHolder(request, sessionHolder)
+
+    // Check if the operation already exists, either in the active execution map, or in the
+    // graveyard of tombstones of executions that have been abandoned. The latter is to prevent
+    // double executions when the client retries, thinking it never reached the server, but in
+    // fact it did, and already got removed as abandoned.
+    if (executions.putIfAbsent(executeHolder.key, executeHolder) != null) {
```

Comment:
You could use computeIfAbsent here... That will allow you to do the tombstone check there.

Reply:
Thanks, also instantiating an ExecuteHolder after checking that it's absent would be more efficient. I'll fix it.

Reply:
-> key is generated within ExecuteHolder, so deferring instantiation won't work. I'll just checking the tombstone part into the lambda.

Reply:
you could move the key generation out :)...